File: Go/Min Stack.go
// 16ms, 93.62%; 7.5MB, 100.00%
type MinStack struct {
	min, stack []int
	mlen, slen int
}

/** initialize your data structure here. */
func Constructor() MinStack {
	return MinStack{stack: make([]int, 0, 1), min: make([]int, 0, 1)}
}

func (this *MinStack) Push(x int) {
	if this.mlen == 0 || this.min[this.mlen-1] >= x {
		this.min = append(this.min, x)
		this.mlen++
	}
	this.stack = append(this.stack, x)
	this.slen++

}

func (this *MinStack) Pop() {
	if this.min[this.mlen-1] == this.stack[this.slen-1] {
		this.min = this.min[:this.mlen-1]
		this.mlen--
	}
	this.stack = this.stack[:this.slen-1]
	this.slen--

}

func (this *MinStack) Top() int {
	return this.stack[this.slen-1]
}

func (this *MinStack) GetMin() int {
	return this.min[this.mlen-1]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: Java/Min Stack.java
// 5ms, 56.64%; 41.3MB, 10.15%
class MinStack {
    private Stack<Integer> st = new Stack<>();
    private Stack<Integer> stMin = new Stack<>();

    public MinStack() {
    }

    public void push(int x) {
        st.addElement(x);
        if (stMin.isEmpty() || stMin.peek() >= x) {
            stMin.addElement(x);
        }
    }

    public void pop() {
        int pop = st.pop();
        if (pop == stMin.peek()) {
            stMin.pop();
        }
    }

    public int top() {
        return st.peek();
    }

    public int getMin() {
        return stMin.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
